pkg/controllers/queue/queue_controller.go

```diff
@@ -0,0 +1,266 @@
+	kblister "github.com/kubernetes-sigs/kube-batch/pkg/client/listers/scheduling/v1alpha1"
+)
+
+// Controller manages queue status.
+type Controller struct {
+	kubeClient kubernetes.Interface
+	kbClient   kbclientset.Interface
+
+	// queueLister
+	queueLister kblister.QueueLister
+	queueSynced cache.InformerSynced
+
+	// podGroup lister
+	pgLister kblister.PodGroupLister
+	pgSynced cache.InformerSynced
+
+	// queues that need to be updated.
+	queue workqueue.RateLimitingInterface
+
+	pgMutex   sync.RWMutex
+	podGroups map[string]map[string]struct{}
+}
+
+func NewQueueController(
+	kubeClient kubernetes.Interface,
+	kbClient kbclientset.Interface,
+) *Controller {
+	factory := kbinformerfactory.NewSharedInformerFactory(kbClient, 0)
+	queueInformer := factory.Scheduling().V1alpha1().Queues()
+	pgInformer := factory.Scheduling().V1alpha1().PodGroups()
+
+	c := &Controller{
+		kubeClient: kubeClient,
+		kbClient:   kbClient,
+
+		queueLister: queueInformer.Lister(),
+		queueSynced: queueInformer.Informer().HasSynced,
+
+		pgLister: pgInformer.Lister(),
+		pgSynced: pgInformer.Informer().HasSynced,
+
+		queue:     workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		podGroups: make(map[string]map[string]struct{}),
+	}
+
+	queueInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addQueue,
+		DeleteFunc: c.deleteQueue,
+	})
+
+	pgInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addPodGroup,
+		UpdateFunc: c.updatePodGroup,
+		DeleteFunc: c.deletePodGroup,
+	})
+
+	return c
+}
+
+// Run starts QueueController
+func (c *Controller) Run(stopCh <-chan struct{}) {
+
+	if !cache.WaitForCacheSync(stopCh, c.queueSynced, c.pgSynced) {
+		glog.Errorf("unable to sync caches for queue controller")
+		return
+	}
+
+	go wait.Until(c.worker, 0, stopCh)
+	glog.Infof("QueueController is running ...... ")
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and
+// marks them done. You may run as many of these in parallel as you wish; the
+// workqueue guarantees that they will not end up processing the same `queue`
+// at the same time.
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	eKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(eKey)
+
+	if err := c.syncQueue(eKey.(string)); err != nil {
+		glog.V(2).Infof("Error syncing queues %q, retrying. Error: %v", eKey, err)
+		c.queue.AddRateLimited(eKey)
+		return true
+	}
+
+	c.queue.Forget(eKey)
+	return true
+}
+
+func (c *Controller) syncQueue(key string) error {
+	glog.V(5).Infoln("Begin sync queue")
+
+	var pending, running, unknown int32
+	c.pgMutex.RLock()
+	podGroups := make([]string, len(c.podGroups[key]))
+	if c.podGroups[key] == nil {
+		glog.V(2).Infoln("queue %s has not been seen or deleted")
+		return nil
+	}
+	for pgKey := range c.podGroups[key] {
```

Comment:
prefer to handle this in Cache: Queue -> Job -> PodGroup.

Reply:
I donot want to overlap with job controller.  For queue controller, it just care about the podgroups, not care about Job or anything else. 

podGroups is keyed by queue name and value is the podgroups that are using this queue.

Reply:
`pkg/cache` is a common component in controller, and we need to clean up jobs when queue was deleted.

Reply:
IC, then we need to record jobs too.

Reply:
But have a doubt: why would user delete their queues before jobs? Is there a use case? If there is a  malicious user who does this, we should not delete the running jobs by any reason.

Reply:
"malicious user" is handled by security; if admin want to delete queue, it's a heavy work to delete all jobs before delete queue.

Reply:
For this, I can think of a more simple way to do this: I remember there will be an admission controller to validate queue existence before create jobs. Then we can also set the job's owner reference.  Then when the queue is deleted, k8s garbage collector can reclaim all children resources automatically. 

So we have two choice:

1. do it in controller, need to record all dependents relationship.

1. setting ownerreference, do nothing else, leave it to GC.

Reply:
Personally prefer not adding this into cache even we need considering deleting jobs when queue deleted.

Reply:
> setting ownerreference, do nothing else, leave it to GC.

owerReference should be reserved for `JobGroup` :)

Reply:
OwnerReferences is an array, can be more than one.  

>owerReference should be reserved for JobGroup :)

Sorry, there is no `JobGroup` now.

Reply:
not sure whether GC will check controlled owner reference.

> there is no JobGroup now.

so "reserved" it.